semantic/poly_types.go

```diff
@@ -822,6 +837,40 @@ func (k ObjectKind) resolvePolyType(kinds map[Tvar]Kind) (PolyType, error) {
 	return NewObjectPolyType(properties, k.lower, k.upper), nil
 }
 
+// NullableKind indicates that it is possible for this
+// variable to be the null type if no other type is
+// more appropriate.
+type NullableKind struct {
+	T PolyType
+}
+
+func (NullableKind) MonoType() (Type, bool)        { return Nil, true }
+func (NullableKind) freeVars(*Constraints) TvarSet { return nil }
+func (n NullableKind) resolveType(kinds map[Tvar]Kind) (Type, error) {
+	return Nil, nil
+}
+func (n NullableKind) resolvePolyType(kinds map[Tvar]Kind) (PolyType, error) {
+	return n.T, nil
+}
+func (n NullableKind) substituteKind(tv Tvar, t PolyType) Kind {
+	if ptv, ok := n.T.(Tvar); ok && ptv == tv {
+		return NullableKind{T: t}
+	}
+	return n
+}
+func (n NullableKind) unifyKind(kinds map[Tvar]Kind, k Kind) (Kind, Substitution, error) {
+	switch k := k.(type) {
+	case ObjectKind:
+		return nil, nil, errors.New("cannot unify null with object kind")
+	case ArrayKind:
+		return nil, nil, errors.New("cannot unify null with array kind")
+	case NullableKind:
+		return k, nil, nil
```

Comment:
Null kinds must unify with other null kinds, because otherwise `a + b` wouldn't type check if both are null correct? Also if we are unifying two nullable kinds, shouldn't we also unify the types they describe like so:
```go
case NullableKind:
    sub, err := unifyTypes(kinds, n.T, k.T)
    if err != nil {
        return nil, nil, err
    }
    return k, sub, nil
```

Reply:
This is where the implementation gets really weird. The kind only applies to `Tvar` and so these methods only get called if the variable is a `Tvar`. If I call `unifyTypes`, then `unifyTypes` consults the kind and this becomes an infinite loop. Objects and arrays don't have to deal with this because they unify their properties rather than themselves.

It seems like this works because when we unify `Tvar` values, it unifies their kinds if they exist so ensuring that both of them have the same kind and therefore a reference to the same `Tvar` guarantees they'll be the same type.

Reply:
Ok I agree. We can't unify the underlying types because it would be an infinite loop. And it's not necessary since if a null kind is being unified with another kind, then it must be true that the underlying types of each are constrained to be equal and will be unified at some point. And so a null kind should only unify with another null kind and there is no need for further recursive unification.